File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        a = b = head
        
        cyc = False
        while b != None and b.next != None:
            a = a.next
            b = b.next.next
            if a == b:
                cyc = True
                break
        if not cyc:
            return None
        c = head
        while cyc and a != c:
            a = a.next
            c = c.next
        return c

File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        ListNode a = head;
        ListNode b = head;
        boolean cyc = false;
        while(b != null && b.next != null){
            a = a.next;
            b = b.next.next;

            if(a == b){
                cyc = true;
                break;
            }

        }
        if(!cyc){
            return null;
        }
        ListNode c = head;
        while(a != c){
            a = a.next;
            c = c.next;
        }
        return c;
    }
}